generator/templates/server/responses.gotmpl

```diff
@@ -28,10 +28,11 @@ if {{ .ValueExpression }}S != "" {
 {{ end }}
 {{define "sliceserverheaderbuilder" }}
 var {{ varname .Child.ValueExpression }}R []string
-for _, {{ varname .Child.ValueExpression }} := range {{ .ValueExpression }} {
-  {{- if not .Child.IsArray }}{{ template "sliceitemserverheaderbuilder" .Child }}{{ else }}{{ template "sliceserverheaderbuilder" .Child }}{{ end -}}
+if len({{ .ValueExpression }}) > 0 {
```

Comment:
doesn't a for loop imply `if len > 0` ?

Reply:
Sure, when it is nested.
But the template is self-reproduced and the first instance and we need to guard against possible ranging into the next level.
Here is an example of code I am currently generating with tests:
Please mark that array outermost validation still go to the validate private func.
For nested arrays, we are compelled to produce some less elegant code.

For param spec:
```yaml
/mixed/deepnested:
    get:
      operationId: getInDepth
      parameters:
      - name: deepNested1
        in: query
        type: array
        collectionFormat: pipes
        required: true
        maxItems: 2
        items:
          type: array
          collectionFormat: csv
          minItems: 2
          items:
            type: array
            collectionFormat: ssv
            uniqueItems: true
            items:
              type: string
              pattern: '^[a-z]$'
```

I got:
```golang
func (o *GetInDepthParams) bindDeepNested1(rawData []string, hasKey bool, formats strfmt.Registry) error {
	if !hasKey {
		return errors.Required("deepNested1", "query")
	}

	var qvDeepNested1 string
	if len(rawData) > 0 {
		qvDeepNested1 = rawData[len(rawData)-1]
	}

	deepNested1IC := swag.SplitByFormat(qvDeepNested1, "pipes")

	if len(deepNested1IC) == 0 {
		return errors.Required("deepNested1", "query")
	}

	var deepNested1IR [][][]string
	for i, deepNested1IV := range deepNested1IC {
		deepNested1IIC := swag.SplitByFormat(deepNested1IV, "csv")
		if len(deepNested1IIC) > 0 {

			var deepNested1IIR [][]string
			for ii, deepNested1IIV := range deepNested1IIC {
				deepNested1IIIC := swag.SplitByFormat(deepNested1IIV, "ssv")
				if len(deepNested1IIIC) > 0 {

					var deepNested1IIIR []string
					for iii, deepNested1IIIV := range deepNested1IIIC {
						deepNested1III := deepNested1IIIV

						if err := validate.Pattern(fmt.Sprintf("%s.%v", fmt.Sprintf("%s.%v", fmt.Sprintf("%s.%v", "deepNested1", i), ii), iii), "query", deepNested1III, `^[a-z]$`); err != nil {
							return err
						}

						deepNested1IIIR = append(deepNested1IIIR, deepNested1III)
					}

					deepNested1IIR = append(deepNested1IIR, deepNested1IIIR)
				}

				// uniqueItems: true
				if err := validate.UniqueItems(fmt.Sprintf("%s.%v", fmt.Sprintf("%s.%v", "deepNested1", i), ii), "query", deepNested1IIIC); err != nil {
					return err
				}

			}

			deepNested1IR = append(deepNested1IR, deepNested1IIR)
		}

		deepNested1ISize := int64(len(deepNested1IIC))

		// minItems: 2
		if err := validate.MinItems(fmt.Sprintf("%s.%v", "deepNested1", i), "query", deepNested1ISize, 2); err != nil {
			return err
		}

	}

	o.DeepNested1 = deepNested1IR
	if err := o.validateDeepNested1(formats); err != nil {
		return err
	}

	return nil
}

func (o *GetInDepthParams) validateDeepNested1(formats strfmt.Registry) error {

	deepNested1Size := int64(len(o.DeepNested1))

	// maxItems: 2
	if err := validate.MaxItems("deepNested1", "query", deepNested1Size, 2); err != nil {
		return err
	}

	return nil
}
```

Reply:
The outer if is unnecessary, as the only thing inside the if is a for loop over the thing that len is testing.  These two blocks have exactly the same meaning.

    for _, {{ varname .Child.ValueExpression }} := range {{ .ValueExpression }} {
    }

and 

    if len({{ .ValueExpression }}) > 0 {
        for _, {{ varname .Child.ValueExpression }} := range {{ .ValueExpression }} {
        }
    }

Reply:
I don't remember how exactly, but I had a case of panic() when testing and added the guard in a hurry. I've to look at it more closely the

Reply:
Ah got it now (it has been a few days...): the if does not only guard the loop but the append() after the loop too. So it's not completely useless.

Reply:
I don't see it.  Do the nested templates do something funky like closing the loop and opening a new one?  I don't see it there either.

    if len({{ .ValueExpression }}) > 0 {
        for _, {{ varname .Child.ValueExpression }} := range {{ .ValueExpression }} {
            {{- if not .Child.IsArray }}
                {{ template "sliceitemserverheaderbuilder" .Child }}
            {{ else }}
                {{ template "sliceserverheaderbuilder" .Child }}
            {{ end -}}
        }
    }

Reply:
BTW hello Greg. And thank you for your comments.

You are looking at the wrong place (this part is to generate headers not parameters - yes, I had to change it too to support arrays, this story shall also be told...).

Look at the sample generated code I provided above. I believe it's more understandable than the template, which is well, ahem, funky, yes.

Reply:
I think you're looking at a different place than we are.  We're looking at a simple `for` loop that was wrapped in an `if` in this PR.  There is no such structure in your generated code above.

https://github.com/fredbi/go-swagger/blob/e77845a4df1c90949bd7a9091ae2344f953d8f2f/generator/templates/server/responses.gotmpl#L31-L35

Reply:
I think you're looking at this, which does contain the `append`:  

https://github.com/fredbi/go-swagger/blob/e77845a4df1c90949bd7a9091ae2344f953d8f2f/generator/templates/server/parameter.gotmpl#L45-L48

Reply:
yeah i am focused on parameters right now. that's the most complex. i have to look again for the response - this one was a quick fix - thanks